### PR TITLE
Dev/issue 8895 file metadata

### PR DESCRIPTION
**Pull request by mcantelon**
I've made the PREMIS object/event parsers without serialization (which could make sense to add in the future to make it easier to create METS files from scratch, but would be a big undertaking).

### CODE REVIEW

File: metsrw/premis_event.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+from . import exceptions
+from . import utils
+
+
+class Event(object):
```
Comment (Hwesta) [inline metsrw/premis_event.py]: Rename to PremisEvent (or PREMISEvent) to distinguish from some other metadata's events?  Or maybe put all the PREMS things in premis.py so you have premis.Event?

File: metsrw/premis_event.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    """
+    An object representing a PREMIS object element.
+
+    :param str event_type: Event type.
+    :param str outcome: Whether an event passed or failed.
+    """
+
+    def __init__(self, event_type=None, outcome=None):
+        self.event_type = event_type
+        self.outcome = outcome
```
Comment (Hwesta) [inline metsrw/premis_event.py]: Events have more attributes than these two.  How will those be handled?  Is there the intention for adding more attributes later? [The PREMIS spec](http://www.loc.gov/standards/premis/v2/premis-2-2.pdf) is actually really readable and clear about what is in it (and optional/required, repeatable etc)

File: metsrw/premis_event.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        """
+        Create a new Event by parsing root.
+
+        :param root: Element or ElementTree to be parsed into an object.
+        :raises exceptions.ParseError: If the root is not event.
+        """
+        if root.tag != utils.lxmlns('premis') + 'event':
+            raise exceptions.ParseError('Object can only parse event elements with PREMIS namespace.')
+
+        event_type_el = root.find(utils.lxmlns('premis') + 'eventType')
```
Comment (Hwesta) [inline metsrw/premis_event.py]: Tip: `.find[|text|all]` takes a namespaces parameter.  eg `root.find('premis:eventType', namespaces=utils.NAMESPACES)`

File: metsrw/premis_object.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+    :param str filename: Filename.
+    :param object characteristics: :class:`ObjectCharacteristics` instance that is part of this PREMIS object.
+    :param str identifier_type: Identifier type.
+    :param str identifier_value: Identifier value.
+    :param str relationship_type: Relationship type.
+    :param str relationship_subtype: Relationship subtype.
+    """
+
+    def __init__(self, filename=None, characteristics=None, identifier_type=None, identifier_value=None, relationship_type=None, relationship_subtype=None):
```
Comment (Hwesta) [inline metsrw/premis_object.py]: Is it a valid object if any or all of these attributes are None? What attributes are required in the PREMIS spec? Repeatable?

File: metsrw/premis_object.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+class ObjectCharacteristics(object):
+    """
+    An object representing a PREMIS objectCharacteristics element.
+
+    :param str size: File size in bytes.
+    :param str format_name: Name of the file format. TODO
+    :param str format_version: Version of the file form.
+    :param str format_registry_name: Name of the format registry.
+    :param str format_registry_key: Registry-specific format key.
+    :param boolean is_mets: Whether or not the file is a METS file.
```
Comment (Hwesta) [inline metsrw/premis_object.py]: What is this required for?

**Comment by Hwesta** [comment]
:sparkles: for the tests.  Add some for error cases?

While writing the PREMIS events/objects back out might not be required, it will make it much clearer what data you need to parse from the elements to be able to write them. The goal is to be able to roundtrip a METS file: read it in, optionally modify, write it out, and get semantically the same METS.  It looks like several of the classes don't store enough data to be able to do this.

File: metsrw/premis/event.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+            raise exceptions.ConstructError('identify_type argument is required.')
+
+        if identifier_value is None:
+            raise exceptions.ConstructError('identify_value argument is required.')
+
+        if event_type is None:
+            raise exceptions.ConstructError('event_type argument is required.')
+
+        if event_datetime is None:
+            raise exceptions.ConstructError('event_datetime argument is required.')
```
Comment (Hwesta) [inline metsrw/premis/event.py]: The easier way to do this is to not provide a default value in the constructor.  ege `def __init__(self, identifier_type, identifier_value, event_type, event_datetime, detail=None, ...)`

File: metsrw/premis/event.py [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+from .. import utils
+
+
+class Event(object):
+    """
+    An object representing a PREMIS object element.
+
+    :param str event_type: Event type.
+    :param str outcomes: Event outcome information.
+    :raises exceptions.ParseError: If the root is not event.
```
Comment (Hwesta) [inline metsrw/premis/event.py]: Docstring needs update

File: metsrw/premis/event.py [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+        self.event_datetime = event_datetime
+        self.detail = detail
+        self.outcomes = outcomes
+        self.linking_agent_identifiers = linking_agent_identifiers
+        self.linking_object_identifiers = linking_object_identifiers
+
+    @classmethod
+    def parse(cls, root):
+        """
+        Create a new Event by parsing root.
```
Comment (Hwesta) [inline metsrw/premis/event.py]: This should probably specify what tag it's expecting - eg premis:event

File: metsrw/utils.py [hunk 9]
```diff
@@ -13,11 +13,20 @@
...
+PREMIS_VERSION = "2.2"
 
 def lxmlns(arg):
     """ Return XPath-usable namespace. """
     return '{' + NAMESPACES[arg] + '}'
 
+def find_element_text(parent_el, path_expression):
+    """ Attempt to find element and, if found, return its text content """
+    el = parent_el.find(path_expression, NAMESPACES)
+    return el.text if el is not None else None
```
Comment (Hwesta) [inline metsrw/utils.py]: This is a duplicate of lxml's findtext. See http://lxml.de/tutorial.html#ElementPath and http://lxml.de/api/index.html under `lxml.etree` -> `ElementBase` -> `findtext`

**Comment by Hwesta** [comment]
Don't forget to run this through a linter (preferably configure your editor to be doing that constantly) and ensure your tests all pass.

File: metsrw/premis_object.py (hunk 5 above)
Comment (mcantelon) [inline metsrw/premis_object.py]: Putting the object characteristics in another class is just a way to break up the functionality into less huge pieces.

**Pull request closed without merging**